iree/compiler/Conversion/LinalgToSPIRV/ConvertToSPIRVPass.cpp

```diff
@@ -126,20 +151,60 @@ Value getPushConstantValue(Operation *op, unsigned elementCount,
   return builder.create<spirv::LoadOp>(loc, acOp);
 }
 
-/// Gets or inserts a resource evariable of the given `type` in `block` and bind
-/// it to `set` and `binding`.
-spirv::GlobalVariableOp getOrInsertResourceVariable(Location loc, Type type,
-                                                    unsigned set,
-                                                    unsigned binding,
-                                                    Block &block) {
-  auto name = llvm::formatv("__resource_var_{0}_{1}__", set, binding).str();
-  for (auto varOp : block.getOps<spirv::GlobalVariableOp>()) {
-    if (varOp.sym_name() == name) return varOp;
+/// Inserts a resource evariable of the given `type` into `block` and bind
+/// it to `set` and `binding`. `id` uniquely identifies the inserted variable.
+spirv::GlobalVariableOp insertResourceVariable(Location loc, Type type,
```

Comment:
Do we need this complicated logic. I would expect that the `getOrInsertResourceVariable` will just add the alias attribute when it finds an existing `spv.globalVariable` with the same set and binding number?

Reply:
I thin that could make performance a bit worse, because we need a way to track spv.globalVariable?

In the past, we are using block.getOps<spirv::GlobalVariableOp>() and it ends with O(n^2) time complexity.

If we track it with DenseMapInfo, it would be O(n log n) time complexity. I feel this way is not that complicated. It more like going through all the placeholder ops and deciding which should attach aliased attribute in the first pass, and create the ops in lowering in the second pass.

I'm okay with both, will leave the decision to you. Just commenting what I think here.

Reply:
Good point Hanhan. We can leave it this way.

Reply:
Thinking about this a bit more. The `getOps` only iterates over the ops in the module which is `spv.globalVariable`s and `spv.function`s. Thats not a lot of operations. I think its OK to go with the `getOps` for the simplicity here.

Reply:
Good point. I was thinking there could be many flow.varaible in the module, but that wouldn't be the case. Because they exists in the outer-module. So `getOps` won't be inefficient in this case.

Both are okay to me, it looks like a preference problem to me. :)

Reply:
IMO, using the existing `getOps` approach to loop through all `spv.globalVariable`s would mean a complicated scheme for uniquing the name strings and managing those strings to figure out which function the global variable is for.

Let's say we have two entry functions: `foo` and `bar`. `foo` has a placeholder pointing to a resource bound at `(x, y)`. `bar` has three placeholders all pointing to `(x, y)`. We'll create a `spv.globalVariable` for `foo` and three `spv.globalVariable`s for `bar`. The name cannot only contain `_x_y_`; otherwise we won't be able to differentiate between `foo` and `bar`. We need to put the function name into it. But then we also need to differentiate between types given all three `spv.globalVariable`s can have different types. So we need a way to encode the type into it too. Using type name is not good; we can probably use the type pointer given types are uniqued. So now we have a very lengthy string for representing the resource: `__resource_var_{set}_{binding}_{function_name}_{type_id}_`. Then when we try to insert, say, the global variable for the third placeholder in `bar`, we need to go over all the existing `spv.globalVariable`s and figure out whether to decorate them as aliased by matching strings.

I feel we are going a bit much with the string based approach so would like a more clean way to handle it. Pre-analyze seems better to me. But I can understand your point here: it does mean some extra C++ code. I prefer the current approach, but not super attached to it. If you have strong preference towards the previous one, I can change back. :)

Reply:
Just so that I understand, why would you need to know if the variable is referenced in `foo` or in `bar`. In either case you only need to know that you are referencing a particular set and binding right? Does it matter which set you are referencing those from? So the name of the resource is `__resource_var_{set}_{binding}_{id}__` with `id` being the pointer value (which is what is done now). So you need to add alias attribute if you have the same set/binding and type. Those you can query from the op right? I dont think you will have a case with different 

There is nothing technically wrong with this approach. Definitely (slightly?) more efficient. I just want to avoid a lot of code complexity that overall decreases readability.

Reply:
> In either case you only need to know that you are referencing a particular set and binding right?

Not really. You need to decorate `Aliased` when two SPIR-V resource variables are bound to the same (set, binding) pair **and** they are used by the same entry function statically. If they are used by two **different** entry points, they are effectively for different shaders/kernels; `Aliased` does not make sense.

Now the question is, does it harm if we just decorate everything using the same (set, binding) pair as `Aliased`, even coming from different entry points? Honestly I don't really know here: it's an area of implementation defined behavior. That probably means something blowing up in the driver compiler. Or if they are okay with it, probably means performance penalty given `Aliased` means now the driver need to make sure that different references need to be synchronized to guarantee correctness. (To be frank, I'm also a bit suspicious whether `Aliased` is currently well supported among different GPU drivers. It's from the spec; but unless something exists in the CTS, otherwise there is a high chance it might break.) ;-P

Reply:
> > In either case you only need to know that you are referencing a particular set and binding right?
> 
> Not really. You need to decorate `Aliased` when two SPIR-V resource variables are bound to the same (set, binding) pair **and** they are used by the same entry function statically. If they are used by two **different** entry points, they are effectively for different shaders/kernels; `Aliased` does not make sense.
> 

That doesnt compute for me. This probably needs to be resolved with  (maybe with David Neto, or someone involved with the spec). If the globalVariable is at module level it doesnt make too much sense to me that whether you add an `Aliased` decoration or not is determined by how they are accessed by multiple entry points within the kernel and not based on whether the two are actual aliases of the underlying `StorageBuffer`. I would think a decent Alias analysis infra would be able to not add unnecessary overhead for things.

> Now the question is, does it harm if we just decorate everything using the same (set, binding) pair as `Aliased`, even coming from different entry points? Honestly I don't really know here: it's an area of implementation defined behavior. That probably means something blowing up in the driver compiler. Or if they are okay with it, probably means performance penalty given `Aliased` means now the driver need to make sure that different references need to be synchronized to guarantee correctness. (To be frank, I'm also a bit suspicious whether `Aliased` is currently well supported among different GPU drivers. It's from the spec; but unless something exists in the CTS, otherwise there is a high chance it might break.) ;-P


I understand driver compilers have different levels of sophistication, but we cannot really be accounting for that. If a driver compiler is having issues handling what is a reasonable decision taken by the SPIR-V backend (and in this case I do believe that the right solution is to add `Aliased` decoration to buffers that actually alias and not predicate it based on where it is accessed from), we should kick the problem back to the driver compiler.

Sorry for being pedantic here, but conflating where things are accessed from to decide whether it is aliased (and the complexity that this seems to be adding) seems unjustified.